src/MMDB.cc

```diff
@@ -0,0 +1,359 @@
+// See the file "COPYING" in the main distribution directory for copyright.
+
+#include "zeek/MMDB.h"
+
+#include <netinet/in.h>
+#include <netinet/ip.h>
+#include <sys/socket.h>
+#include <sys/stat.h>
+#include <chrono>
+
+#include "zeek/Func.h"
+#include "zeek/IPAddr.h"
+#include "zeek/ZeekString.h"
+
+namespace zeek {
+
+#ifdef USE_GEOIP
+
+static int msg_count = 0;
+static double msg_suppression_time = 0;
+static constexpr int msg_limit = 20;
+static constexpr double msg_suppression_duration = 300;
+
+LocDB mmdb_loc;
+AsnDB mmdb_asn;
+
+static void report_msg(const char* format, ...) {
+    if ( zeek::run_state::network_time > msg_suppression_time + msg_suppression_duration ) {
+        msg_count = 0;
+        msg_suppression_time = zeek::run_state::network_time;
+    }
+
+    if ( msg_count >= msg_limit )
+        return;
+
+    ++msg_count;
+
+    va_list al;
+    va_start(al, format);
+    std::string msg = zeek::util::vfmt(format, al);
+    va_end(al);
+
+    zeek::reporter->Info("%s", msg.data());
+}
+
+static zeek::ValPtr mmdb_getvalue(MMDB_entry_data_s* entry_data, int status, int data_type) {
+    switch ( status ) {
+        case MMDB_SUCCESS:
+            if ( entry_data->has_data ) {
+                switch ( data_type ) {
+                    case MMDB_DATA_TYPE_UTF8_STRING:
+                        return zeek::make_intrusive<zeek::StringVal>(entry_data->data_size, entry_data->utf8_string);
+                        break;
+
+                    case MMDB_DATA_TYPE_DOUBLE:
+                        return zeek::make_intrusive<zeek::DoubleVal>(entry_data->double_value);
+                        break;
+
+                    case MMDB_DATA_TYPE_UINT32: return zeek::val_mgr->Count(entry_data->uint32);
+
+                    default: break;
+                }
+            }
+            break;
+
+        case MMDB_LOOKUP_PATH_DOES_NOT_MATCH_DATA_ERROR:
+            // key doesn't exist, nothing to do
+            break;
+
+        default: report_msg("MaxMind DB error [%s]", MMDB_strerror(status)); break;
+    }
+
+    return nullptr;
+}
+
+MMDB::MMDB() : mmdb{}, file_info{}, reported_error{false}, last_check{zeek::run_state::network_time} {}
+
+MMDB::~MMDB() { Close(); }
+
+bool MMDB::OpenFile(const char* a_filename) {
+    filename = a_filename;
+    Close();
+
+    if ( 0 != stat(a_filename, &file_info) ) {
+        return false;
+    }
+
+    int status = MMDB_open(a_filename, MMDB_MODE_MMAP, &mmdb);
```

Comment:
```suggestion
    int status = MMDB_open(filename.c_str(), MMDB_MODE_MMAP, &mmdb);
```

Nit, but I was a bit wondering if a_filename is dangling (but must have been in the prior implementation).

Reply:
Yeah. I did a quick test and the library does copy the string. But why risk it — I changed it to `filename`, thanks.